runtime/compiler/optimizer/InterpreterEmulator.cpp

```diff
@@ -0,0 +1,849 @@
+   if (createTargetStack)
+      _stacks[targetIndex] = new (comp()->trStackMemory()) ByteCodeStack(comp()->trMemory(), std::max<uint32_t>(20, _stack->size()));
+   }
+
+void
+InterpreterEmulator::initializeIteratorWithState()
+   {
+   _iteratorWithState = true;
+   _unknownOperand = new (comp()->trStackMemory()) Operand();
+   uint32_t size = this->maxByteCodeIndex() + 5;
+   _flags  = (flags8_t *) this->trMemory()->allocateStackMemory(size * sizeof(flags8_t));
+   _stacks = (ByteCodeStack * *) this->trMemory()->allocateStackMemory(size * sizeof(ByteCodeStack *));
+   memset(_flags, 0, size * sizeof(flags8_t));
+   memset(_stacks, 0, size * sizeof(ByteCodeStack *));
+   _stack = new (comp()->trStackMemory()) TR_Stack<Operand *>(comp()->trMemory(), 20, false, stackAlloc);
+
+   genBBStart(0);
+   setupBBStartContext(0);
+   this->setIndex(0);
+   }
+
+void
+InterpreterEmulator::maintainStack(TR_J9ByteCode bc)
+   {
+   TR_ASSERT_FATAL(_iteratorWithState, "has to be called when the iterator has state!");
+   int slotIndex = -1;
+   switch (bc)
+      {
+      case J9BCgetfield: maintainStackForGetField(); break;
+      case J9BCaload0: slotIndex = 0; maintainStackForAload(slotIndex); break;
+      case J9BCaload1: slotIndex = 1; maintainStackForAload(slotIndex); break;
+      case J9BCaload2: slotIndex = 2; maintainStackForAload(slotIndex); break;
+      case J9BCaload3: slotIndex = 3; maintainStackForAload(slotIndex); break;
+      case J9BCaload:  slotIndex = nextByte(); maintainStackForAload(slotIndex); break;
+      case J9BCaloadw: slotIndex = next2Bytes(); maintainStackForAload(slotIndex); break;
+
+      case J9BCinvokespecial:
+      case J9BCinvokespecialsplit:
+         maintainStackForDirectCall(_calltarget->_calleeMethod);
+         break;
+      case J9BCiconstm1: push (new (comp()->trStackMemory()) IconstOperand(-1)); break;
+      case J9BCiconst0:  push (new (comp()->trStackMemory()) IconstOperand(0)); break;
+      case J9BCiconst1:  push (new (comp()->trStackMemory()) IconstOperand(1)); break;
+      case J9BCiconst2:  push (new (comp()->trStackMemory()) IconstOperand(2)); break;
+      case J9BCiconst3:  push (new (comp()->trStackMemory()) IconstOperand(3)); break;
+      case J9BCiconst4:  push (new (comp()->trStackMemory()) IconstOperand(4)); break;
+      case J9BCiconst5:  push (new (comp()->trStackMemory()) IconstOperand(5)); break;
+      case J9BCifne:
+         push (new (comp()->trStackMemory()) IconstOperand(0));
+         maintainStackForIf(J9BCificmpne);
+         break;
+      case J9BCifeq:
+         push (new (comp()->trStackMemory()) IconstOperand(0));
+         maintainStackForIf(J9BCificmpeq);
+         break;
+      case J9BCgoto:
+         genTarget(bcIndex() + next2BytesSigned());
+         break;
+      case J9BCpop:
+      case J9BCputfield:
+      case J9BCputstatic:
+      case J9BCladd:
+      case J9BCiadd:
+      case J9BCisub:
+         pop();
```

Comment:
this seems a bit wrong - the add/sub opcodes do two pops and the pop/putfield/putstatic do one - can you explain why I don't see two pops?

Reply:
I was using one `pop` because there would be 2 pop and one push for the add/sub opcodes but now I realize I have to do the 2 pops and push an unknown operand. Changed to code accordingly.